### PR TITLE
Better NaN/inf loss handling for O0 (skip step across workers)

### PR DESCRIPTION
**Pull request by redoctopus**
`actions.py` now syncs across workers to check for NaN/inf loss. Terminates all workers if `stop_on_nan_loss` is set, ignores it and lets apex deal with it if amp optimization level is O1 or higher, and skips the step across workers otherwise.

Signed-off-by: Jocelyn Huang <jocelynh@nvidia.com>

### CODE REVIEW

File: nemo/backends/pytorch/actions.py [hunk 1]
```diff
@@ -1418,26 +1399,22 @@ def train(
...
-                                self.step,
-                            )
-                        else:
-                            logging.warning('Loss is NaN or inf, continuing training')
                     final_loss += registered_tensors[tensor.unique_name]
 
+                # Check for NaN/inf loss (across workers if applicable)
+                loss_nan_inf_checker = final_loss.clone()
+                if placement_gpu:
+                    dist.all_reduce(loss_nan_inf_checker)
```
Comment (blisc) [inline nemo/backends/pytorch/actions.py:1407]: Isn't an overflow possible here? Since the default is SUM.

What is the reasoning to check every step?

Comment (redoctopus) [reply nemo/backends/pytorch/actions.py]: Good catch, apparently I forgot that floats overflow to inf. Changed this to use MAX.

We should check every step to prevent NaN/inf loss from ever propagating any garbage gradient at any step.
